content/en/blog/_posts/2022-08-16-PSP-historical-context/index.md

```diff
@@ -0,0 +1,131 @@
+layout: blog
+title: "PodSecurityPolicy: The Historical Context"
+date: 2022-08-16
+slug: podsecuritypolicy-the-historical-context
+---
+
+**Author:** Mahé Tardy (Quarkslab)
+
+The PodSecurityPolicy (PSP) admission controller will be removed as of
+Kubernetes v1.25. Its deprecation was announced and detailed in the blog post
+[PodSecurityPolicy Deprecation: Past, Present, and Future](/blog/2021/04/06/podsecuritypolicy-deprecation-past-present-and-future/),
+published for the Kubernetes v1.21 release.
+
+This article aims at providing historical context on the birth and evolution of
+PSP, to explain why the feature never made it to stable and had to be removed
+in favor of a new in-tree admission controller. Let's recap what
+PodSecurityPolicy is, why it was necessary, the historical context of its
+addition to Kubernetes, and finally, why it was removed and replaced by Pod
+Security admission control.
+
+PodSecurityPolicy, like other specialized admission control plugins, provided
+fine-grained permissions on specific fields concerning the pod security
+settings as a built-in policy API. It acknowledged the fact that cluster
+administrators and cluster users are, most of the time, not the same people, and
+that creating workloads in the form of a Pod or any resource that will
+create a Pod should not equal being "root on the cluster". PodSecurityPolicy
+could also encourage best practices by configuring more secure defaults through
+mutation and decoupling low-level Linux security decisions from the deployment
+process.
+
+## The birth of PodSecurityPolicy
+
+PodSecurityPolicy originated from OpenShift's SecurityContextConstraints
+(SCC) that were in the very first release of the Red Hat container platform,
+even before Kubernetes 1.0. PSP was a stripped-down version of the SCC.
+
+<!-- not sure the next paragraph is relevant -->
+
+The origin of the creation of PSP is difficult to track notably because it was
+mainly added before Kubernetes Enhancements Proposal (KEP) process, when design
+proposals were still a thing. Indeed, the archive of the final
+[design proposal](https://github.com/kubernetes/design-proposals-archive/blob/main/auth/pod-security-policy.md)
+is still available. Nevertheless, a KEP issue was created after the first pull
+requests were merged, and is [KEP number five](https://github.com/kubernetes/enhancements/issues/5).
+
+Before adding the first piece of code that created the PSP, two main pull
+requests were merged into Kubernetes, [the security context subresource](https://github.com/kubernetes/kubernetes/pull/7343)
+that defined the new fields on pods' containers and [the service account](https://github.com/kubernetes/kubernetes/pull/7101)
+feature. Then, Kubernetes 1.0 was released on 10 July 2015 without a mechanism
+to restrict the security context and sensitive options of workloads, apart from
+scdeny, an admission control that just prevents usage of some fields in the
+security context, which still exists today.
```

Comment:
```suggestion
feature. Then, apart from the scdeny package, Kubernetes 1.0 was released on 10 July 2015 without 
a mechanism to restrict the security context and sensitive options of workloads.
The scdeny package is still in use today and creates an admission controller that prevents the usage of some fields in the security context.
```

Reply:
You are perfectly correct, it's [technically a package](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/securitycontext/scdeny/admission.go) but it feels a bit weird to talk about an admission plugin as a package, it seems a bit too generic!
```suggestion
feature. Then, apart from the scdeny admission plugin, Kubernetes 1.0 was released on 10 July 2015 without 
a mechanism to restrict the security context and sensitive options of workloads.
The scdeny plugin is still in use today and creates an admission controller that prevents the usage of some fields in the security context.
```
What do you think of that?

Reply:
Looks great to me.